### PR TITLE
Use QgsTask instead of QThread

### PR DESCRIPTION
**Pull request by ptormene**
Once the thread is started, the OQ-Consolidate dialog is closed and the progress of the consolidation is shown in a progress bar in the bottom part of the QGIS GUI. If you click on the progress bar, you can see the running threads (one, in this case) and you can abort the thread. Notifications are displayed in the QGIS message bar on top of the QGIS GUI.

### CODE REVIEW

**Review by daniviga — APPROVED** [review]
LGTM as user :+1:

**Review by daniviga — APPROVED** [review]
LGTM

**Comment by mbernasocchi** [comment]
checking

File: consolidate_task.py [hunk 1]
```diff
@@ -20,62 +20,60 @@
...
+class ConsolidateTask(QgsTask):
 
-        self.iface = iface
+    def __init__(self, description, flags, outputDir, projectFile, saveToZip):
+        super().__init__(description, flags)
         self.outputDir = outputDir
         self.layersDir = outputDir + "/layers"
         self.projectFile = projectFile
         self.saveToZip = saveToZip
+        self.progressRange = None
```
Comment (mbernasocchi) [inline consolidate_task.py]: maybe progressMax

File: consolidate_task.py [hunk 2]
```diff
@@ -92,81 +90,63 @@ def consolidate(self):
...
+            lType = layer.type()
+            lProviderType = layer.providerType()
+            lName = layer.name()
+            lUri = layer.dataProvider().dataSourceUri()
+            if lType == QgsMapLayer.VectorLayer:
+                # Always convert to GeoPackage
+                outFile = self.convertGenericVectorLayer(
+                    e, layer, lName)
+                outFiles.append(outFile)
+            elif lType == QgsMapLayer.RasterLayer:
```
Comment (mbernasocchi) [inline consolidate_task.py]: why don't we put this also in the geopackage?

File: consolidate_task.py [hunk 3]
```diff
@@ -92,81 +90,63 @@ def consolidate(self):
...
+        return True
 
     def loadProject(self):
         f = QFile(self.projectFile)
         if not f.open(QIODevice.ReadOnly | QIODevice.Text):
             msg = self.tr("Cannot read file %s:\n%s.") % (self.projectFile,
                                                           f.errorString())
-            self.processError.emit(msg)
-            return
+            raise Exception(msg)
```
Comment (mbernasocchi) [inline consolidate_task.py]: raise IOException?

File: consolidate_task.py [hunk 4]
```diff
@@ -92,81 +90,63 @@ def consolidate(self):
...
 
         doc = QDomDocument()
         setOk, errorString, errorLine, errorColumn = doc.setContent(f, True)
         if not setOk:
             msg = (self.tr("Parse error at line %d, column %d:\n%s")
                    % (errorLine, errorColumn, errorString))
-            self.processError.emit(msg)
-            return
+            raise Exception(msg)
```
Comment (mbernasocchi) [inline consolidate_task.py]: raise ParseException?

File: consolidate_task.py [hunk 5]
```diff
@@ -176,8 +156,7 @@ def saveProject(self, doc):
         if not f.open(QIODevice.WriteOnly | QIODevice.Text):
             msg = self.tr("Cannot write file %s:\n%s.") % (self.projectFile,
                                                            f.errorString())
-            self.processError.emit(msg)
-            return
+            raise Exception(msg)
```
Comment (mbernasocchi) [inline consolidate_task.py]: raise IOException?

File: consolidate_task.py [hunk 6]
```diff
@@ -189,23 +168,26 @@ def zipfiles(self, file_paths, archive):
         :param file_paths: list of path names
         :param archive: path of the archive
         """
+        if self.isCanceled():
+            raise Exception('Consolidation canceled')
```
Comment (mbernasocchi) [inline consolidate_task.py]: raise TaskCancelledException?

File: consolidate_task.py [hunk 7]
```diff
@@ -189,23 +168,26 @@ def zipfiles(self, file_paths, archive):
...
             os.path.commonprefix([os.path.dirname(f) for f in file_paths]))
         with zipfile.ZipFile(
                 archive, 'w', zipfile.ZIP_DEFLATED, allowZip64=True) as z:
-            for f in file_paths:
+            for i, f in enumerate(file_paths):
                 z.write(f, f[prefix:])
-                self.updateProgress.emit()
+                self.setProgress(i / self.progressRange * 100)
+                if self.isCanceled():
+                    raise Exception('Consolidation canceled')
```
Comment (mbernasocchi) [inline consolidate_task.py]: raise TaskCancelledException?

File: consolidate_task.py [hunk 8]
```diff
@@ -189,23 +168,26 @@ def zipfiles(self, file_paths, archive):
...
 
     def copyXmlRasterLayer(self, layerElement, vLayer, layerName):
         outFile = "%s/%s.xml" % (self.layersDir, layerName)
         try:
             copyfile(vLayer.dataProvider().dataSourceUri(), outFile)
         except IOError:
             msg = self.tr("Cannot copy layer %s") % layerName
-            self.processError.emit(msg)
-            return
+            raise Exception(msg)
```
Comment (mbernasocchi) [inline consolidate_task.py]: raise IOError?

File: consolidate_task.py [hunk 9]
```diff
@@ -229,8 +211,7 @@ def convertGenericVectorLayer(self, layerElement, vLayer, layerName):
             vLayer, outFile, enc, crs, 'GPKG')
         if error != QgsVectorFileWriter.NoError:
             msg = self.tr("Cannot copy layer %s: %s") % (layerName, error_msg)
-            self.processError.emit(msg)
-            return
+            raise Exception(msg)
```
Comment (mbernasocchi) [inline consolidate_task.py]: raise IOException?

File: oq_consolidate.pro [hunk 10]
```diff
@@ -0,0 +1,5 @@
+SOURCES = __init__.py \
```
Comment (mbernasocchi) [inline oq_consolidate.pro]: do we use the pro file? it is usually for QtCreator

File: oq_consolidate.py [hunk 11]
```diff
@@ -31,14 +31,14 @@
...
-from . import aboutdialog
+from . import oq_consolidate_dialog
+from . import about_dialog
 from .utils import log_msg, tr
 
 from . import resources_rc  # NOQA
 
 
-class QConsolidatePlugin(object):
+class OQConsolidatePlugin(object):
```
Comment (mbernasocchi) [inline oq_consolidate.py]: In general we should consider making this plugin QConsolidate and then add on top of it add the little specificity that we have ad OQConsolidate

**Review by mbernasocchi — CHANGES_REQUESTED** [review]
All looks good, just the very generic exceptions are not really cool. Also, In general we should consider making this plugin QConsolidate and then add on top of it add the little specificity that we have ad OQConsolidate

File: consolidate_task.py (hunk 2 above)
Comment (ptormene) [reply consolidate_task.py]: I've added a FIXME about the need to check this

Comment (daniviga) [reply consolidate_task.py]: be careful that here we are not managing any raster layer, but only the ones defined as GDAL virtual layers to expose TMS as rasters in the map, using an external XML. This part isn't needed anymore and we should instead extend it to any raster

**Review by daniviga — APPROVED** [review]
LGTM

**Pull request merged**